consul/src/main/scala/io/buoyant/consul/v1.scala

```diff
@@ -121,6 +126,128 @@ package object v1 {
+
+    val kvPrefix = s"$uriPrefix/kv"
+
+    private[this] val infiniteRetryFilter = new RetryFilter[http.Request, http.Response](
+      RetryPolicy.backoff(backoffs) {
+        // We will assume 5xx are retryable, everything else is not for now
+        case (_, Return(rep)) => rep.status.code >= 500 && rep.status.code < 600
+        case (_, Throw(NonFatal(ex))) =>
+          log.error(s"retrying consul kv request on error $ex")
+          true
+      },
+      HighResTimer.Default,
+      stats,
+      RetryBudget.Infinite
+    )
+
+    def retryClient(retry: Boolean) = {
+      val retryFilter = if (retry) infiniteRetryFilter else Filter.identity[http.Request, http.Response]
+      retryFilter andThen client
+    }
+
+    // https://www.consul.io/docs/agent/http/kv.html#single
+    def list(
+      path: String,
+      datacenter: Option[String] = None,
+      blockingIndex: Option[String] = None,
+      separator: Option[String] = Some("/"),
+      retry: Boolean = false
+    ): Future[Indexed[Seq[String]]] = {
+      val req = mkreq(
+        http.Method.Get,
+        s"$kvPrefix$path",
+        "keys" -> Some(true.toString),
+        "separator" -> separator,
+        "index" -> blockingIndex,
+        "dc" -> datacenter
+      )
+      Trace.letClear(retryClient(retry)(req)).flatMap {
+        case rsp if rsp.status == http.Status.NotFound => Future.exception(NotFound(rsp))
+        case rsp => Future.const(Indexed.mk[Seq[String]](rsp))
+      }
+    }
+
+    // https://www.consul.io/docs/agent/http/kv.html#single
+    def get(
+      path: String,
+      datacenter: Option[String] = None,
+      blockingIndex: Option[String] = None,
+      retry: Boolean = false
+    ): Future[Indexed[String]] = {
+      val req = mkreq(
+        http.Method.Get,
+        s"$kvPrefix$path",
+        "raw" -> Some(true.toString),
+        "index" -> blockingIndex,
+        "dc" -> datacenter
+      )
+
+      Trace.letClear(retryClient(retry)(req)) flatMap {
+        case rsp if rsp.status == http.Status.Ok =>
+          Future.value(Indexed[String](rsp.getContentString(), rsp.headerMap.get(Headers.Index)))
+        case rsp if rsp.status == http.Status.NotFound => Future.exception(NotFound(rsp))
+        case rsp if rsp.status == http.Status.Forbidden => Future.exception(Forbidden(rsp))
+        case rsp => Future.exception(UnexpectedResponse(rsp))
+
+      }
+    }
+
+    // https://www.consul.io/docs/agent/http/kv.html#single
+    def put(
+      path: String,
+      value: String,
+      datacenter: Option[String] = None,
+      cas: Option[String] = None,
+      retry: Boolean = false
+    ): Future[Boolean] = {
+      val req = mkreq(
+        http.Method.Put,
+        s"$kvPrefix$path",
+        "cas" -> cas,
+        "dc" -> datacenter
+      )
+      req.setContentString(value)
+      Trace.letClear(retryClient(retry)(req)).flatMap {
+        case rsp if rsp.status == http.Status.Ok =>
+          Future.const(readJson[Boolean](rsp.content))
```

Comment:
Could use `Indexed.mk` here too.  Of course, the return type of this method would change to `Future[Indexed[Boolean]]` and the index would always be `None`, but I think that's okay.